designs/0034-jacobian.md

```diff
@@ -0,0 +1,320 @@
+- Feature Name: `jacobian` target
+- Start Date: 01-25-2024
+
+# Summary
+[summary]: #summary
+
+This design doc proposes adding a `jacobian` target and a new block for user defined constraints. The `jacobian` target will be accessible directly in `transformed parameters` and a new `constraints` block where users can define custom constraints similar to Stan's already existing constrained data types. The below examples shows an example use case.
+
+```stan
+constraints {
+ upper_bound {
+   real constrain(real x, real upper_bound) {
+     jacobian += x;
+     return upper_bound - exp(x);
+   }
+   real constrain(vector x, real upper_bound) {
+     jacobian += sum(x);
+     return upper_bound - exp(x);
+   }
+   real unconstrain(real x, real upper_bound) {
+     return log(upper_bound - x);
+   }
+   // Used for data
+   int validate(real x, real upper_bound) {
+     // For most this will be
+     return x < upper_bound;
+   }
+ }
+}
+data {
+ real lb;
+ real ub;
+ int N;
+}
+parameters {
+ //(2) User defined constraint
+ real<udc=upper_bound(ub)> b;
```

Comment:
I like `constrain` a lot more!

Reply:
I also prefer `constraint` but I'd leave the equals sign out so it's just
```
real<constraint upper_bound(ub)> b;
```
Otherwise it looks too much like a (complete) function call. The syntax should make it obvious the following is _not_ allowed
```
tmp = upper_bound(ub);
real<constraint=tmp> b;
```

Reply:
The example
```
tmp = upper_bound(ub);
real<constraint=tmp> b;
```
Would be trivially disallowed by the inability to give a type to `tmp`

That said, I also kind of like omitting the `=`

Reply:
I also like omitting the `=`!